### ISSUE TITLE
Geth sets limit for open files too low after startup, results in "too many open files"

### ISSUE THREAD

**Issue by tantrum-junkie** [ethereum/go-ethereum#16006]
I guess that might be related to how the file limit is set by this function:
https://github.com/ethereum/go-ethereum/blob/fdb34b7a7c87142cde48a2ec092f77a56d66f773/common/fdlimit/fdlimit_unix.go#L25

Observed behaviour:
=================
```
lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 17.10
Release:        17.10
Codename:       artful
```

```
./geth --rpc --rpcport 8102 --rpcaddr 192.168.0.1 --cache=2048 --maxpeers=10
INFO [02-01|09:29:56] Starting peer-to-peer node               instance=Geth/v1.7.3-stable-4bb3c89d/linux-amd64/go1.9.2
INFO [02-01|09:29:56] Allocated cache and file handles         database=/home/miner/.ethereum/geth/chaindata cache=2048 handles=1024
INFO [02-01|09:29:59] Initialised chain configuration          config="{ChainID: 1 Homestead: 1150000 DAO: 1920000 DAOSupport: true EIP150: 2463000 EIP155: 2675
000 EIP158: 2675000 Byzantium: 4370000 Engine: ethash}"
INFO [02-01|09:29:59] Disk storage enabled for ethash caches   dir=/home/miner/.ethereum/geth/ethash count=3
INFO [02-01|09:29:59] Disk storage enabled for ethash DAGs     dir=/home/miner/.ethash               count=2
INFO [02-01|09:29:59] Initialising Ethereum protocol           versions="[63 62]" network=1
INFO [02-01|09:29:59] Loaded most recent local header          number=5006992 hash=c37de0…ddc325 td=2303374863991927018140
INFO [02-01|09:29:59] Loaded most recent local full block      number=0       hash=d4e567…cb8fa3 td=17179869184
INFO [02-01|09:29:59] Loaded most recent local fast block      number=5006744 hash=749e77…bfba64 td=2302708533829412214126
INFO [02-01|09:29:59] Upgrading chain index                    type=bloombits percentage=1
WARN [02-01|09:29:59] Failed to load transaction journal       err="open /home/miner/.ethereum/geth/transactions.rlp: too many open files"
WARN [02-01|09:29:59] Failed to rotate transaction journal     err="open /home/miner/.ethereum/geth/transactions.rlp.new: too many open files"
ERROR[02-01|09:29:59] Section processing failed                type=bloombits error="canonical block #77824 unknown"
INFO [02-01|09:29:59] Starting P2P networking
Fatal: Error starting protocol stack: listen udp :30303: socket: too many open files
```

On a second console I'm polling the ulimits for the geth process with
```cat /proc/$(pidof geth)/limits | grep files```

Directly after the start this reports:
```Max open files            150000               150000               files```

(which is what I configured for my system)

But shortly after that the limits seem to get reduced by the geth process itself to
```Max open files            2048               150000               files```

which then results in the "too many open files" problem.

Workaround:
==========
Running this command continously in a seperate shell during geth startup raises the ulimits back to the intended value fast enough for geth to properly start up and begin syncing
```prlimit -n150000 -p $(pidof geth)```

So far this seems to fix it.

Expected behaviour:
==========
Do not crash with "too many open files" ;)

Proposed fix:
==========
Set the correct max open files from within geth or don't do it at all and let the OS manage it.

Hope that helps someone, that is also struggeling with this problem.
And maybe this is an easy fix for the devs.
Thanks for your time.

**Comment by holiman** [comment]
Hm, interesting. In https://github.com/ethereum/go-ethereum/blob/master/common/fdlimit/fdlimit_unix.go, it says
```
// Raise tries to maximize the file descriptor allowance of this process
// to the maximum hard-limit allowed by the OS.
func Raise(max uint64) error {
``` 
But in reality, it does _not_ try to set the allowance to the max allowed by the OS, but instead caps it to the `max` value provided to `Raise`. This is either an error in the docs, or implementation. I'd prefer to let the documentation remain and change the behaviour to not do anything if the os-allowance is larger than `max`.

**Comment by holiman** [comment]
Specifically: https://github.com/ethereum/go-ethereum/blob/c335821479db9930a98cbd48996f880c35a59797/cmd/utils/flags.go#L716 : 
```golang
func makeDatabaseHandles() int {
	if err := fdlimit.Raise(2048); err != nil {
		Fatalf("Failed to raise file descriptor allowance: %v", err)
	}
	limit, err := fdlimit.Current()
```

Should be
```golang
func makeDatabaseHandles() int {
	limit, err := getFdLimit()
	if err != nil {
		Fatalf("Failed to retrieve file descriptor allowance: %v", err)
	}
	if limit < 2048 {
		if err := raiseFdLimit(2048); err != nil{
			Fatalf("Failed to raise file descriptor allowance: %v", err)
		}
	}
```

**Comment by tantrum-junkie** [comment]
Ah, nice.
That seems to be the source of the "2048" value.
As far as I understand your fix should work as desired and not reduce a much higher OS limit to 2048 anymore.

But since we are at it:
As far as I can see at the end of this function it seems that half of the available file handles are reserved for "networking and other stuff":
https://github.com/ethereum/go-ethereum/blob/c335821479db9930a98cbd48996f880c35a59797/cmd/utils/flags.go#L727

Maybe this is also a bit excessive?

Thanks for responding so fast and finding a probable solution to the problem.

**Comment by holiman** [comment]
I think the idea is to limit the database handles, and I'm guessing it's not to make the database go crazy with too many opened files.

**Comment by tantrum-junkie** [comment]
Unfortunately I do not know nearly enough about the inner workings of the geth client to make any informed suggestion, but I can obeserve on my system that whne starting the geth client it seems to open nearly every file in the chaindata directory:

```
lsof -p $(pidof geth) | wc -l
25953
```

```
ls -al ~/.ethereum/geth/chaindata/ | wc -l
26115
```
At the moment when I executed the above commands the client was in the process of "Upgrading chain index".

If I understand it right. I need to have at least a ulimit setting for open files of double the number of files in the chaindata directory so that I don't get the "too many open files" error.

Just trying to understand how it all fits together ;)

Again. Thanks for taking the time to fix this!

**Comment by nutmix** [comment]
This bug is still there in 1.8.13-stable-225171a4.

We are plagued by "accept: Too many open files" whenever we try to us the rpc apis. 

We tried raising the proces and user limits to 32,000, but it always is limited to around 2000.

```
cat /proc/sys/fs/file-nr
2112	0	3280354
```

```
cat /proc/2160/limits | grep files
Max open files            32000                32000                files     
```

**Comment by nutmix** [comment]
This bug is still there in 1.8.13-stable-225171a4.

We are plagued by "accept: Too many open files" whenever we try to us the rpc apis.

We tried raising the proces and user limits to 32,000, but it always is limited to around 2000.

cat /proc/sys/fs/file-nr
2112	0	3280354
cat /proc/2160/limits | grep files
Max open files            32000                32000                files

**Comment by abednarek** [comment]
I've been gathering data on a local node using batches of async requests (getBalance,GetTransferCount). On my dev machine (Win10) i've been sending out async queries in batches of 10,000 at a time which take about 8 seconds to complete. No problems there. 

I ran the same code in the cloud on an Ubuntu 18.04 distro and have been getting tons of "missing trie node" errors on getBalance, GetTransferCount. Reducing batch size requests from 10,000 to 100 fixed the issue. After digging around I came across here and tantrum-junkie workaround of bumping up the file limit (prlimit -n150000 -p $(pidof geth)) fixed the problem and I was able to make async requests in batch sizes of 10,000 again.

I think when the open file limit is hit the error message of "missing trie node" is misleading. Maybe a file open error should be caught with an appropriate message returned?
